File: rb/eyes-selenium/spec/applitools/selenium/spec-driver_spec.rb
require('selenium-webdriver')
require('webdrivers/chromedriver')
require_relative('../../../lib/applitools/selenium/spec-driver')
require_relative('../../../lib/applitools/refer')
require_relative('../../spec_helper')

describe 'spec-driver' do
  before(:all) do
    options = ::Selenium::WebDriver::Chrome::Options.new
    options.add_argument('--headless')
    @driver = ::Selenium::WebDriver.for :chrome, options: options
    @url = 'https://applitools.github.io/demo/TestPages/FramesTestPage/'
    @driver.get @url
  end
  after(:all) do
    @driver.quit
  end
  it('isElement(element)') do
    el = @driver.find_element(:css, 'div')
    expect(::Applitools::Selenium::SpecDriver.isElement(el)).to eq(true)
  end
  it('isEqualElements(element, element)') do
    el1 = @driver.find_element(:css, 'div')
    expect(::Applitools::Selenium::SpecDriver.isEqualElements(@driver, el1, el1)).to eq(true)
  end
  it('isEqualElements(element1, element2)') do
    el1 = @driver.find_element(:css, 'div')
    el2 = @driver.find_element(:css, 'h1')
    expect(::Applitools::Selenium::SpecDriver.isEqualElements(@driver, el1, el2)).to eq(false)
  end
  it('executeScript') do
    args = [0, 'string', {key: 'value'}, [0, 1, 2, 3]]
    expected = @driver.execute_script('return arguments', args)
    result = ::Applitools::Selenium::SpecDriver.executeScript(@driver, 'return arguments', args)
    expect(result).to eq(expected.first)
  end
  it('executeScript error handling') do
    expect { ::Applitools::Selenium::SpecDriver.executeScript(@driver, 'console.log(a.a)') }.to raise_error(Selenium::WebDriver::Error::JavascriptError)
  end
  it('mainContext') do
    begin
      mainDocument = @driver.find_element(:css, 'html')
      @driver.switch_to.frame(@driver.find_element(:css, '[name="frame1"]'))
      @driver.switch_to.frame(@driver.find_element(:css, '[name="frame1-1"]'))
      frameDocument = @driver.find_element(:css, 'html')
      expect(::Applitools::Selenium::SpecDriver.isEqualElements(@driver, mainDocument, frameDocument)).to eq(false)
      ::Applitools::Selenium::SpecDriver.mainContext(@driver)
      resultDocument = @driver.find_element(:css, 'html')
      expect(::Applitools::Selenium::SpecDriver.isEqualElements(@driver, resultDocument, mainDocument)).to eq(true)
    ensure
      @driver.switch_to.default_content
    end
  end
  it('parentContext') do
    begin
      @driver.switch_to.frame(@driver.find_element(:css, '[name="frame1"]'))
      parentDocument = @driver.find_element(:css, 'html')
      @driver.switch_to.frame(@driver.find_element(:css, '[name="frame1-1"]'))
      frameDocument = @driver.find_element(:css, 'html')
      expect(::Applitools::Selenium::SpecDriver.isEqualElements(@driver, parentDocument, frameDocument)).to eq(false)
      ::Applitools::Selenium::SpecDriver.parentContext(@driver)
      resultDocument = @driver.find_element(:css, 'html')
      expect(::Applitools::Selenium::SpecDriver.isEqualElements(@driver, resultDocument, parentDocument)).to eq(true)
    ensure
      @driver.switch_to.default_content
    end
  end
  it('childContext') do
    begin
      element = @driver.find_element(:css, '[name="frame1"]')
      @driver.switch_to.frame(element)
      expectedDocument = @driver.find_element(:css, 'html')
      @driver.switch_to.default_content
      ::Applitools::Selenium::SpecDriver.childContext(@driver, element)
      resultDocument = @driver.find_element(:css, 'html')
      expect(::Applitools::Selenium::SpecDriver.isEqualElements(@driver, resultDocument, expectedDocument)).to eq(true)
    ensure
      @driver.switch_to.default_content
    end
  end
  it('findElement css') do
    expected = @driver.find_element(:css, 'div')
    actual = ::Applitools::Selenium::SpecDriver.findElement(@driver, 'div')
    expect(::Applitools::Selenium::SpecDriver.isEqualElements(@driver, expected, actual)).to eq(true)
  end
  it('findElement eyes-selectors') do
    expected = @driver.find_element(:css, 'div')
    actual = ::Applitools::Selenium::SpecDriver.findElement(@driver, {type: 'css', selector: 'div'})
    expect(::Applitools::Selenium::SpecDriver.isEqualElements(@driver, expected, actual)).to eq(true)
  end
  it('findElement xpath') do
    skip
  end
  it('findElements') do
    expected = @driver.find_elements(:css, 'div')
    actual = ::Applitools::Selenium::SpecDriver.findElements(@driver, 'div')
    expect(::Applitools::Selenium::SpecDriver.isEqualElements(@driver, expected, actual)).to eq(true)
  end
  it('findElements eyes-selectors') do
    expected = @driver.find_elements(:css, 'div')
    actual = ::Applitools::Selenium::SpecDriver.findElements(@driver, {type: 'css', selector: 'div'})
    expect(::Applitools::Selenium::SpecDriver.isEqualElements(@driver, expected, actual)).to eq(true)
  end
  it('getTitle') do
    expected = @driver.title
    result = ::Applitools::Selenium::SpecDriver.getTitle(@driver)
    expect(result).to eq(expected)
  end
  it('getUrl') do
    result = ::Applitools::Selenium::SpecDriver.getUrl(@driver)
    expect(result).to eq(@url)
  end
  it('isMobile') do
    result = ::Applitools::Selenium::SpecDriver.getDriverInfo(@driver)
    expect(result[:isMobile]).to eq(false)
  end
  it('getPlatformName') do
    result = ::Applitools::Selenium::SpecDriver.getDriverInfo(@driver)
    expect(result[:platformName]).to_not be_nil
  end
  it('takeScreenshot') do
    expected = @driver.screenshot_as(:base64)
    actual = ::Applitools::Selenium::SpecDriver.takeScreenshot(@driver)
    expect(actual).to eq(expected)
  end
end

describe('onscreen desktop') do
  before(:all) do
    options = Selenium::WebDriver::Chrome::Options.new
    @driver = Selenium::WebDriver.for :chrome, options: options
    @driver.get 'https://applitools.github.io/demo/TestPages/FramesTestPage/'
  end
  after(:all) do
    @driver.quit
  end
  it('getWindowRect') do
    rect = @driver.manage.window.rect.to_h
    result = ::Applitools::Selenium::SpecDriver.getWindowRect(@driver)
    expect(result).to eq(rect)
  end
  it('setWindowRect') do
    input = {x: 50, y: 50, width: 510, height: 511}
    ::Applitools::Selenium::SpecDriver.setWindowRect(@driver, input)
    rect = @driver.manage.window.rect
    expect(rect.height).to eq(input[:height])
    expect(rect.width).to eq(input[:width])
    expect(rect.x).to eq(input[:x])
    expect(rect.y).to eq(input[:y]) 
  end
end

#describe('mobile driver (@mobile)', async () => {
#  before(async () => {
#    ;[driver, destroyDriver] = await spec.build({browser: 'chrome', device: 'Pixel 3a XL'})
#  })
#
#  after(async () => {
#    await destroyDriver()
#  })
#
#  it('isMobile()', isMobile({expected: true}))
#  it('getDeviceName()', getDeviceName({expected: 'Google Pixel 3a XL GoogleAPI Emulator'}))
#  it('getPlatformName()', getPlatformName({expected: 'Android'}))
#  it('isNative()', isNative({expected: false}))
#  it('getOrientation()', getOrientation({expected: 'portrait'}))
#  it('getPlatformVersion()', getPlatformVersion({expected: '10'}))
#})

File: packages/spec-driver-selenium/test/e2e/spec-driver.spec.ts
import assert from 'assert'
import {By} from 'selenium-webdriver'
import * as spec from '../../src'

describe('spec driver', async () => {
  let driver: spec.Driver, destroyDriver: () => void
  const url = 'https://applitools.github.io/demo/TestPages/FramesTestPage/'

  describe('headless desktop', async () => {
    before(async () => {
      ;[driver, destroyDriver] = await spec.build({browser: 'chrome'})
      driver = spec.transformDriver(driver)
      await driver.get(url)
    })

    after(async () => {
      await destroyDriver()
    })

    it('isDriver(driver)', async () => {
      await isDriver({input: driver, expected: true})
    })
    it('isDriver(wrong)', async () => {
      await isDriver({input: {} as spec.Driver, expected: false})
    })
    it('isElement(element)', async () => {
      await isElement({input: await driver.findElement({css: 'div'}), expected: true})
    })
    it('isElement(wrong)', async () => {
      await isElement({input: {} as spec.Element, expected: false})
    })
    it('isSelector(by)', async () => {
      await isSelector({input: By.xpath('//div'), expected: true})
    })
    it('isSelector(by-hash)', async () => {
      await isSelector({input: {xpath: '//div'}, expected: true})
    })
    it('isSelector(wrong)', async () => {
      await isSelector({input: {} as spec.Selector, expected: false})
    })
    it('transformSelector(by)', async () => {
      await transformSelector({input: By.css('div'), expected: By.css('div')})
    })
    it('transformSelector(string)', async () => {
      await transformSelector({input: '.element', expected: {css: '.element'}})
    })
    it('transformSelector(common-selector)', async () => {
      await transformSelector({input: {selector: '.element', type: 'css'}, expected: {css: '.element'}})
    })
    it('isEqualElements(element, element)', async () => {
      const element = await driver.findElement({css: 'div'})
      await isEqualElements({input: {element1: element, element2: element}, expected: true})
    })
    it('isEqualElements(element1, element2)', async () => {
      await isEqualElements({
        input: {element1: await driver.findElement({css: 'div'}), element2: await driver.findElement({css: 'h1'})},
        expected: false,
      })
    })
    it('executeScript(strings, args)', async () => {
      await executeScript()
    })
    it('findElement(selector)', async () => {
      await findElement({input: {selector: {css: '#overflowing-div'}}})
    })
    it('findElement(selector, parent-element)', async () => {
      await findElement({input: {selector: {css: 'div'}, parent: await driver.findElement({css: '#stretched'})}})
    })
    it('findElement(non-existent)', async () => {
      await findElement({input: {selector: {css: 'non-existent'}}, expected: null})
    })
    it('findElements(string)', async () => {
      await findElements({input: {selector: {css: 'div'}}})
    })
    it('findElements(string, parent-element)', async () => {
      await findElements({input: {selector: {css: 'div'}, parent: await driver.findElement({css: '#stretched'})}})
    })
    it('findElements(non-existent)', async () => {
      await findElements({input: {selector: {css: 'non-existent'}}, expected: []})
    })
    it('mainContext()', async () => {
      await mainContext()
    })
    it('parentContext()', async () => {
      await parentContext()
    })
    it('childContext(element)', async () => {
      await childContext()
    })
    it('getTitle()', async () => {
      await getTitle()
    })
    it('getUrl()', async () => {
      await getUrl()
    })
    it('visit()', async () => {
      await visit()
    })
    it('getDriverInfo()', async () => {
      await getDriverInfo({
        expected: {
          browserName: 'chrome',
          isMobile: false,
          isNative: false,
          platformName: 'linux',
        },
      })
    })
    it('getWindowSize()', async () => {
      await getWindowSize()
    })
    it('setWindowSize({width, height})', async () => {
      await setWindowSize({input: {width: 551, height: 552}})
    })
  })

  describe('legacy driver (@webdriver)', async () => {
    before(async () => {
      ;[driver, destroyDriver] = await spec.build({browser: 'ie-11', legacy: true})
    })

    after(async () => {
      await destroyDriver()
    })

    it('getWindowSize()', async () => {
      await getWindowSize()
    })
    it('setWindowSize({width, height})', async () => {
      await setWindowSize({input: {width: 551, height: 552}})
    })
    it('getDriverInfo()', async () => {
      await getDriverInfo({
        expected: {
          browserName: 'internet explorer',
          browserVersion: '11',
          isMobile: false,
          isNative: false,
          platformName: 'WINDOWS',
        },
      })
    })
  })

  describe('mobile driver (@mobile @android)', async () => {
    before(async () => {
      ;[driver, destroyDriver] = await spec.build({browser: 'chrome', device: 'Pixel 3a XL'})
      driver = spec.transformDriver(driver)
    })

    after(async () => {
      await destroyDriver()
    })

    it('getWindowSize()', async () => {
      await getWindowSize()
    })
    it('getOrientation()', async () => {
      await getOrientation({expected: 'portrait'})
    })
    it('getDriverInfo()', async () => {
      await getDriverInfo({
        expected: {
          browserName: 'chrome',
          deviceName: 'Google Pixel 3a XL GoogleAPI Emulator',
          isMobile: true,
          isNative: false,
          platformName: 'Android',
          platformVersion: '10',
        },
      })
    })
  })

  describe('mobile driver (@mobile @native @android)', async () => {
    before(async () => {
      ;[driver, destroyDriver] = await spec.build({
        app: 'https://applitools.jfrog.io/artifactory/Examples/android/1.3/app-debug.apk',
        device: 'Pixel 3a XL',
        orientation: 'landscape',
      })
      driver = spec.transformDriver(driver)
    })

    after(async () => {
      await destroyDriver()
    })

    it('getWindowSize()', async () => {
      await getWindowSize()
    })
    it('getOrientation()', async () => {
      await getOrientation({expected: 'landscape'})
    })
    it('getDriverInfo()', async () => {
      await getDriverInfo({
        expected: {
          deviceName: 'Google Pixel 3a XL GoogleAPI Emulator',
          isMobile: true,
          isNative: true,
          platformName: 'Android',
          platformVersion: '10',
          pixelRatio: 2.5,
          statusBarHeight: 60,
          navigationBarHeight: 1080,
        },
      })
    })
  })

  async function isDriver({input, expected}: {input: spec.Driver; expected: boolean}) {
    const result = await spec.isDriver(input || driver)
    assert.strictEqual(result, expected)
  }
  async function isElement({input, expected}: {input: spec.Element; expected: boolean}) {
    const result = await spec.isElement(input)
    assert.strictEqual(result, expected)
  }
  async function isSelector({input, expected}: {input: spec.Selector; expected: boolean}) {
    const result = await spec.isSelector(input)
    assert.strictEqual(result, expected)
  }
  async function transformSelector({input, expected}: {input: any; expected: spec.Selector}) {
    const result = spec.transformSelector(input)
    assert.deepStrictEqual(result, expected || input)
  }
  async function isEqualElements({
    input,
    expected,
  }: {
    input: {element1: spec.Element; element2: spec.Element}
    expected: boolean
  }) {
    const result = await spec.isEqualElements(driver, input.element1, input.element2)
    assert.deepStrictEqual(result, expected)
  }
  async function executeScript() {
    const args = [0, 'string', {key: 'value'}, [0, 1, 2, 3]]
    const expected = await driver.executeScript('return arguments[0]', args)
    const result = await spec.executeScript(driver, 'return arguments[0]', args)
    assert.deepStrictEqual(result, expected)
  }
  async function mainContext() {
    try {
      const mainDocument = await driver.findElement(By.css('html'))
      await driver.switchTo().frame(await driver.findElement(By.css('[name="frame1"]')))
      await driver.switchTo().frame(await driver.findElement(By.css('[name="frame1-1"]')))
      const frameDocument = await driver.findElement(By.css('html'))
      assert.ok(!(await spec.isEqualElements(driver, mainDocument, frameDocument)))
      await spec.mainContext(driver)
      const resultDocument = await driver.findElement(By.css('html'))
      assert.ok(await spec.isEqualElements(driver, resultDocument, mainDocument))
    } finally {
      await driver
        .switchTo()
        .defaultContent()
        .catch(() => null)
    }
  }
  async function parentContext() {
    try {
      await driver.switchTo().frame(await driver.findElement(By.css('[name="frame1"]')))
      const parentDocument = await driver.findElement(By.css('html'))
      await driver.switchTo().frame(await driver.findElement(By.css('[name="frame1-1"]')))
      const frameDocument = await driver.findElement(By.css('html'))
      assert.ok(!(await spec.isEqualElements(driver, parentDocument, frameDocument)))
      await spec.parentContext(driver)
      const resultDocument = await driver.findElement(By.css('html'))
      assert.ok(await spec.isEqualElements(driver, resultDocument, parentDocument))
    } finally {
      await driver
        .switchTo()
        .frame(null)
        .catch(() => null)
    }
  }
  async function childContext() {
    try {
      const element = await driver.findElement(By.css('[name="frame1"]'))
      await driver.switchTo().frame(element)
      const expectedDocument = await driver.findElement(By.css('html'))
      await driver.switchTo().frame(null)
      await spec.childContext(driver, element)
      const resultDocument = await driver.findElement(By.css('html'))
      assert.ok(await spec.isEqualElements(driver, resultDocument, expectedDocument))
    } finally {
      await driver
        .switchTo()
        .frame(null)
        .catch(() => null)
    }
  }
  async function findElement({
    input,
    expected,
  }: {
    input: {selector: spec.Selector; parent?: spec.Element}
    expected?: spec.Element
  }) {
    const root = input.parent ?? driver
    expected = expected === undefined ? await root.findElement(input.selector) : expected
    const element = await spec.findElement(driver, input.selector, input.parent)
    if (element !== expected) {
      assert.ok(await spec.isEqualElements(driver, element, expected))
    }
  }
  async function findElements({
    input,
    expected,
  }: {
    input: {selector: spec.Selector; parent?: spec.Element}
    expected?: spec.Element[]
  }) {
    const root = input.parent ?? driver
    expected = expected === undefined ? await root.findElements(input.selector) : expected
    const elements = await spec.findElements(driver, input.selector, input.parent)
    assert.strictEqual(elements.length, expected.length)
    for (const [index, element] of elements.entries()) {
      assert.ok(await spec.isEqualElements(driver, element, expected[index]))
    }
  }
  async function getWindowSize() {
    let expected
    const window = driver.manage().window()
    if (window.getSize) {
      expected = await window.getSize()
    } else {
      const {width, height} = await window.getRect()
      expected = {width, height}
    }
    const result = await spec.getWindowSize(driver)
    assert.deepStrictEqual(result, expected)
  }
  async function setWindowSize({input}: {input: {width: number; height: number}}) {
    await spec.setWindowSize(driver, input)
    let result
    const window = driver.manage().window()
    if (window.getSize) {
      const {x, y} = await window.getPosition()
      const {width, height} = await window.getSize()
      result = {x, y, width, height}
    } else {
      result = await window.getRect()
    }
    assert.deepStrictEqual(result, {x: 0, y: 0, ...input})
  }
  async function getOrientation({expected}: {expected: 'portrait' | 'landscape'}) {
    const result = await spec.getOrientation(driver)
    assert.strictEqual(result, expected)
  }
  async function getTitle() {
    const expected = await driver.getTitle()
    const result = await spec.getTitle(driver)
    assert.deepStrictEqual(result, expected)
  }
  async function getUrl() {
    const result = await spec.getUrl(driver)
    assert.deepStrictEqual(result, url)
  }
  async function visit() {
    const blank = 'about:blank'
    await spec.visit(driver, blank)
    const actual = await driver.getCurrentUrl()
    assert.deepStrictEqual(actual, blank)
    await driver.get(url)
  }
  async function getDriverInfo({expected}: {expected: Record<string, any>}) {
    const info = await spec.getDriverInfo(driver)
    assert.deepStrictEqual(
      Object.keys(expected).reduce((obj, key) => ({...obj, [key]: info[key]}), {}),
      expected,
    )
  }
})
